### PR TITLE
Add cancel_all_calls to Python server

### PR DESCRIPTION
**Pull request by soltanmm**
cc #2952

### CODE REVIEW

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py [hunk 1]
```diff
@@ -211,5 +211,76 @@ def testEcho(self):
...
+    client_call = self.client_channel.create_call(self.client_completion_queue, METHOD, HOST, DEADLINE)
+    client_start_batch_result = client_call.start_batch([
+        _types.OpArgs.send_initial_metadata([]),
+        _types.OpArgs.send_message(REQUEST, 0),
+        _types.OpArgs.send_close_from_client(),
+        _types.OpArgs.recv_initial_metadata(),
+        _types.OpArgs.recv_message(),
+        _types.OpArgs.recv_status_on_client()
+    ], client_call_tag)
+    self.assertEquals(_types.CallError.OK, client_start_batch_result)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Always leave the optional ess off the end of "assertEqual".

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py [hunk 2]
```diff
@@ -211,5 +211,76 @@ def testEcho(self):
...
+    DEADLINE_TOLERANCE = 0.25
+    CANCEL_ALL_CALLS_TIME_TOLERANCE = 0.5
+    REQUEST = 'in death a member of project mayhem has a name'
+    METHOD = 'twinkies'
+    HOST = 'hostess'
+    server_request_tag = object()
+    request_call_result = self.server.request_call(self.server_completion_queue, server_request_tag)
+
+    client_call_tag = object()
+    client_call = self.client_channel.create_call(self.client_completion_queue, METHOD, HOST, DEADLINE)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Line-wrap just after opening parenthesis to stay within eighty columns?

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py [hunk 3]
```diff
@@ -211,5 +211,76 @@ def testEcho(self):
...
+        _types.OpArgs.send_initial_metadata([]),
+        _types.OpArgs.send_message(REQUEST, 0),
+        _types.OpArgs.send_close_from_client(),
+        _types.OpArgs.recv_initial_metadata(),
+        _types.OpArgs.recv_message(),
+        _types.OpArgs.recv_status_on_client()
+    ], client_call_tag)
+    self.assertEquals(_types.CallError.OK, client_start_batch_result)
+
+    client_no_event, request_event, = wait_for_events([self.client_completion_queue, self.server_completion_queue], time.time() + 2)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Eighty columns.

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py (hunk 1 above)
Comment (soltanmm) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Actually, this was an oversight - that assert shouldn't even be there (the relevant assertions are in the test above this one).

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py [hunk 4]
```diff
@@ -211,5 +211,76 @@ def testEcho(self):
...
+    self.client_completion_queue.shutdown()
+    while self.client_completion_queue.next().type != _types.EventType.QUEUE_SHUTDOWN:
+      pass
+
+    del self.client_completion_queue
+    del self.server_completion_queue
+    del self.server
+
+  def testHangingServerCall(self):
+    DEADLINE = time.time()+5
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Function-scope and method-scope constants are written as normal variables (all_lowercase_with_underscores).

Hey, I didn't write the rules...

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py (hunk 2 above)
Comment (soltanmm) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Are we finally approaching the point where we start caring about line-wrapping? :-P Sure.

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py [hunk 5]
```diff
@@ -211,5 +211,76 @@ def testEcho(self):
...
+  def setUp(self):
+    self.server_completion_queue = _low.CompletionQueue()
+    self.server = _low.Server(self.server_completion_queue, [])
+    self.port = self.server.add_http2_port('[::]:0')
+    self.client_completion_queue = _low.CompletionQueue()
+    self.client_channel = _low.Channel('localhost:%d'%self.port, [])
+
+    self.server.start()
+
+  def tearDown(self):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: This tearDown method doesn't match its corresponding setUp - a setUp/tearDown pair should always be able to function in the presence of someone inserting an empty testUpDown method into the test, but in this case self.server_completion_queue is shut down during the test method.

But hey, it's idempotent, so I think that that means that you could if you wanted to shut it down in tearDown as well as in the test?

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py (hunk 1 above)
Comment (soltanmm) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Oh, and an aside, `CallError` might be an enum, but because we're interfacing with C I made it an `IntEnum`, so `is` doesn't really make much sense when we care about the value rather than the reference.

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py [hunk 6]
```diff
@@ -211,5 +211,76 @@ def testEcho(self):
...
+
+    # Now try to shutdown the server and expect that we see server shutdown
+    # almost immediately after calling cancel_all_calls.
+    with self.assertRaises(RuntimeError):
+      self.server.cancel_all_calls()
+    shutdown_tag = object()
+    self.server.shutdown(shutdown_tag)
+    pre_cancel_timestamp = time.time()
+    self.server.cancel_all_calls()
+    self.server_completion_queue.shutdown()
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Shutting down the server's completion queue seems outside the scope of the test - shouldn't the behavior under test function as we expect it to function without distinction between the completion queue remaining active versus being shut down?

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py (hunk 4 above)
Comment (soltanmm) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Yeeeeeep. I'll reformat the previous test suite as well, then, I guess.

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py (hunk 5 above)
Comment (soltanmm) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Technically... yes?

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py [hunk 7]
```diff
@@ -211,5 +211,76 @@ def testEcho(self):
...
+    finish_shutdown_timestamp = None
+    found_shutdown_tag = False
+    while True:
+      event = self.server_completion_queue.next()
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: If we passed a timeout value in this .next call, we could get around having to shut down the queue, right?

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py [hunk 8]
```diff
@@ -211,5 +211,76 @@ def testEcho(self):
...
+
+  def testHangingServerCall(self):
+    DEADLINE = time.time()+5
+    DEADLINE_TOLERANCE = 0.25
+    CANCEL_ALL_CALLS_TIME_TOLERANCE = 0.5
+    REQUEST = 'in death a member of project mayhem has a name'
+    METHOD = 'twinkies'
+    HOST = 'hostess'
+    server_request_tag = object()
+    request_call_result = self.server.request_call(self.server_completion_queue, server_request_tag)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: Blank line between test data (expressed as constants at the top of the test method; thank you for that) and the first statement of test behavior (actively interacting with the system under test).

File: src/python/grpcio_test/grpc_test/_adapter/_low_test.py (hunk 7 above)
Comment (soltanmm) [inline src/python/grpcio_test/grpc_test/_adapter/_low_test.py]: True.

**Comment by soltanmm** [comment]
Updated and rebased s.t. idempotency is a feature of the source tree (fixed recently in #2991); PTAL.

**Comment by nathanielmanistaatgoogle** [comment]
Looks good; waiting on test results to load...

**Comment by nathanielmanistaatgoogle** [comment]
Jenkins is green for Python; merging.
